### ISSUE TITLE
Creating a new Driver for Mail

### ISSUE THREAD

**Issue by ghost** [adonisjs/adonis-framework#272]
Hi,

I've followed the tutorial in the official documentation for creating a new driver to the mail provider, but instead of PostMark I need to use `nodemailer-sparkpost-transport`, and just adapted some code given in this tutorial. Issue is that I always receive an error saying `Unable to locate sparkpost mail driver`, and I noticed the docs does not specify a folder for the driver class. Question is: where should I put the `SparkPost` driver class? I tried inside of `app/Listeners/`, but still same error when extending the class with `Ioc`, inside of `Http.js`

Much appreciated.

**Comment by thetutlage** [comment]
Can u share some code on how you are extending the Mail driver?

**Comment by mfgabriel92** [comment]
Allow me to share everything.

First, the `SparkPost` class:

```
'use strict'

/**
 * SparkPost driver for sending e-mails
 *
 * @author gabrielmf
 * @class
 */
class SparkPost {

  /**
   * Constructor
   *
   * @author gabrielmf
   */
  constructor(Config) {
    this.config = Config
    this.transport = this._createTransport('mail.sparkpost')
  }

  /**
   * Create Transport 
   *
   * @author gabrielmf
   */
  _createTransport(configKey) {
    let options = this.config.get(configKey)
    let nodemailer = require('nodemailer')
    let sparkpostTransport = require('nodemailer-sparkpost-transport')

    return nodemailer.createTransport(sparkpostTransport(options))
  }

  /**
   * Send email function 
   *
   * @author gabrielmf
   */
  send(message, configKey) {
    let transport = this.transport

    if (configKey) {
      transport = this._createTransport(configKey)
    }
    
    return transport.sendMail(message)
  }
}

module.exports = SparkPost
```

In `mail.js`, I created this:

```
  sparkpost: {
    sparkPostApiKey: Env.get('SPARKPOST_API_KEY')
  },
```

Now, `Http.js`:

```
Ioc.extend('Mail', 'sparkpost', function(app) {
    let Config = app.use('Adonis/Src/Config')

    return new SparkPost(Config)
  })
```

I have a `MailOperation` which makes use of the newly created class:

```
'use strict'

const Mail = use('Mail')

/**
 * Email operations class
 *
 * @author gabrielmf
 * @class
 */
class MailOperation {

  /**
   * Send email 
   *
   * @author gabrielmf
   * @param {String} to the email address which will receive the email
   */
  * send() {
      yield Mail.driver('sparkpost').send('emails.confirm', (message) => {
        message.to('mfgabriel92@gmail.com', 'Gabriel')
        message.from('accounts@fufulabs.com')
        message.subject('Test')
      })
  }
}

module.exports = MailOperation
```

The error is here, saying it cannot find module `sparkpost`. I've put `SparkPost` and `Http` classes together in `Listeners`

**Comment by thetutlage** [comment]
@mfgabriel92 This is my bad, I have updated the docs to be accurate. But lemme share the issues.

1. The extend process needs to be happen before loading the app. So `bootstrap/extend.js` is the file for that.
2. You have to use the actual `namespace` instead of the alias, since before booting the app aliases have not been registered.

So it has to be

#### bootstrap/extend.js

```javascript
Ioc.extend('Adonis/Addons/Mail', 'sparkpost', function(app) {
    let Config = app.use('Adonis/Src/Config')

    return new SparkPost(Config)
})
```

**Comment by mfgabriel92** [comment]
Hi,

I'm glad I could help to improve the documentation. I've followed your instructions there, but I'm now receiving the error `callback must be a function`. That's it. I don't know where to look, for it's just like the examples. I tried to log somethings to understand the flow, but sadly I have no clue. Could you possible have forgotten something else due to these changes?

**Comment by mfgabriel92** [comment]
Hi,

Thank you, I'm glad to help with the improvement of the documentation by pointing out errors. I've followed what you wrote but now there is another error `callback must be a function`. The `MailOperation` class implements this method:

```
  ...
  * send() {
      yield Mail.driver('sparkpost').send('emails.confirm', (message) => {
        message.from('<email>', '<name>')
        message.to('<email>', '<name>')
        message.subject('Test')
      })
  }
```
I've logged somethings to understand the flow but I have no clue. Is it possible to have something missing in the documentation?

**Comment by mfgabriel92** [comment]
Hi,

Thank you, it works now. One more thing I think I can help with the documentation. Before I was receiving the error `callback must be a function`, and I realized when calling the newly created method should be like this:

```
yield Mail.driver('sparkpost').send('emails.confirm', {}, (message) => { ... }
```

I based myself on the `send` method in `adonis-mail-provider/src/Mail/MailManager/index.js:93`

```
* send (view, data, callback, config) {
    if (typeof (callback) !== 'function') {
      throw new NE.InvalidArgumentException('callback must be function')
    }
    // ...
```

Again, thank you.

**Comment by thetutlage** [comment]
Fixed in 3.1 docs http://adonisjs.com/docs/3.1/mail#_adding_new_drivers
